src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+                    .lower_bound_slope = current_metadata.lower_bound_slope,
+                    .upper_bound_slope = current_metadata.upper_bound_slope,
+                };
+                try timestamps_array.append(current_metadata.start_time);
+            }
+        }
+
+        if (timestamps_array.items.len > 0) {
+            // Append the final merged segment metadata.
+            for (timestamps_array.items) |timestamp| {
+                try merged_segments_metadata.append(.{
+                    .start_time = timestamp,
+                    .interception = merge_metadata.interception,
+                    .lower_bound_slope = merge_metadata.lower_bound_slope,
+                    .upper_bound_slope = merge_metadata.upper_bound_slope,
+                });
+            }
+            timestamps_array.clearRetainingCapacity();
+        }
+    }
+
+    // This step is needed since the timestamp order is lost due to the hashmap.
+    mem.sort(
+        SegmentMetadata,
+        merged_segments_metadata.items,
+        {},
+        compareMetadataByStartTime(),
+    );
+}
+
+/// Sim-Piece Phase 3. Compute a hash map from interception points in `merged_segments_metadata`
+/// to a hash map from the approximation slope to an array list of timestamps and store in
+/// `merged_segments_metadata_map`. The `allocator` is used to allocate memory of intermediates.
+fn computeSegmentsMetadataHashMap(
+    merged_segments_metadata: ArrayList(SegmentMetadata),
+    merged_segments_metadata_map: *HashMap(f64, HashMap(
+        f64,
+        ArrayList(usize),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ), HashF64Context, std.hash_map.default_max_load_percentage),
+    allocator: mem.Allocator,
+) !void {
+    for (merged_segments_metadata.items) |segment_metadata| {
+        const interception: f64 = segment_metadata.interception;
+        const slope: f64 = (segment_metadata.lower_bound_slope +
+            segment_metadata.upper_bound_slope) / 2;
+
+        // Get or put the outer HashMap entry for the given interception
+        const hash_to_hash_result = try merged_segments_metadata_map.getOrPut(interception);
+        if (!hash_to_hash_result.found_existing) {
+            hash_to_hash_result.value_ptr.* = HashMap(
+                f64,
+                ArrayList(usize),
+                HashF64Context,
+                std.hash_map.default_max_load_percentage,
+            ).init(allocator);
+        }
+        // Get or put the inner HashMap entry for the given slope
+        const hash_to_array_result = try hash_to_hash_result.value_ptr.*.getOrPut(slope);
+        if (!hash_to_array_result.found_existing) {
+            hash_to_array_result.value_ptr.* = ArrayList(usize).init(allocator);
+        }
+        try hash_to_array_result.value_ptr.*.append(segment_metadata.start_time);
+    }
+}
+
+/// Sim-Piece Phase 4. Create from the `merged_segments_metadata_map` and byte array that can be
```

Comment:
Maybe add another sentence or two stating what the compressed representation is?

Reply:
Fixed.